src/iterative_ensemble_smoother/sies.py

```diff
@@ -0,0 +1,416 @@
+            self.d.reshape(-1, 1)
+            + sample_mvnormal(
+                C_dd_cholesky=self.C_dd_cholesky, rng=self.rng, size=self.X.shape[1]
+            )
+        ).astype(parameters.dtype, copy=False)
+
+        self.W = np.zeros(
+            shape=(self.X.shape[1], self.X.shape[1]), dtype=parameters.dtype
+        )
+
+    def evaluate_objective(
+        self, W: npt.NDArray[np.double], Y: npt.NDArray[np.double]
+    ) -> npt.NDArray[np.double]:
+        """Evaluate the objective function in Equation (18), taking the mean
+        over each ensemble member.
+
+        Parameters
+        ----------
+        W : npt.NDArray[np.double]
+            Current weight matrix W, which represents the current X_i as a linear
+            combination of the prior. See equation (17) and line 9 in Algorithm 1.
+        Y : npt.NDArray[np.double]
+            Reponses when evaluating the model at X_i. In other words, Y = g(X_i).
+
+        Returns
+        -------
+        float
+            The objective function value. Lower objective is better.
+
+        """
+        (prior, likelihood) = self.evaluate_objective_elementwise(W, Y)
+        answer: npt.NDArray[np.double] = (prior + likelihood).mean()
+        return answer
+
+    def evaluate_objective_elementwise(
+        self, W: npt.NDArray[np.double], Y: npt.NDArray[np.double]
+    ) -> tuple[npt.NDArray[np.double], npt.NDArray[np.double]]:
+        """Equation (18) elementwise and termwise, returning a tuple of vectors
+        (prior, likelihood).
+
+        Parameters
+        ----------
+        W : npt.NDArray[np.double]
+            Current weight matrix W, which represents the current X_i as a linear
+            combination of the prior. See equation (17) and line 9 in Algorithm 1.
+        Y : npt.NDArray[np.double]
+            Reponses when evaluating the model at X_i. In other words, Y = g(X_i).
+
+        Returns
+        -------
+        prior : npt.NDArray[np.double]
+            A 1D array representing distance from prior for each ensemble member.
+        likelihood : npt.NDArray[np.double]
+            A 1D array representing distance from observations for each ensemble member.
+
+        """
+        # Evaluate the elementwise prior term
+        # Evaluate np.diag(W.T @ W) = (W**2).sum(axis=0)
+        prior = (W**2).sum(axis=0)
+
+        if self.C_dd.ndim == 2:
+            # Evaluate the elementwise likelihood term
+            # Evaluate np.diag((g(X_i) - D).T @ C_dd^{-1} @ (g(X_i) - D))
+            #        = np.diag((Y - D).T @ C_dd^{-1} @ (Y - D))
+            # First let A := Y - D, then use the cholesky factors C_dd = L @ L.T
+            # to write (Y - D).T @ C_dd^{-1} @ (Y - D)
+            #             A.T @ (L @ L.T)^-1 @ A
+            #             (L^-1 @ A).T @ (L^-1 @ A)
+            # To compute the expression above, we solve K := L^-1 @ A,
+            # then we compute np.diag(K.T @ K) as (K**2).sum(axis=0)
+            A = Y - self.D
```

Comment:
Can we use a different symbol than `A` here?
Also, `A = Y-self.D` can be moved above the `if` since it's used in both `if` and `else`.